nessie/system_test.go

```diff
@@ -0,0 +1,154 @@
+package nessie
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	genclient "github.com/treeverse/lakefs/api/gen/client"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/thanhpk/randstr"
+	"github.com/treeverse/lakefs/api"
+	"github.com/treeverse/lakefs/api/gen/client/setup"
+	"github.com/treeverse/lakefs/api/gen/models"
+	"github.com/treeverse/lakefs/logging"
+)
+
+// testsConfig contains configuration needed for running the tests
+type testsConfig struct {
+	// baseURL is the base address of the lakeFS endpoint
+	baseURL string
+
+	// rawBucketPath is the full path to the s3 bucket
+	rawBucketPath string
+
+	enabled bool
+	https   bool
+}
+
+var (
+	config testsConfig
+	logger logging.Logger
+	client api.Client
+)
+
+const (
+	maxSetup     = 5 * time.Minute
+	masterBranch = "master"
+)
+
+func init() {
+	flag.StringVar(&config.baseURL, "endpoint-url", "localhost:8000", "URL endpoint of the lakeFS instance")
+	flag.StringVar(&config.rawBucketPath, "bucket", "s3://nessie-system-testing", "Bucket's path")
+	flag.BoolVar(&config.enabled, "system-tests", false, "Enabling the system-tests")
+	flag.BoolVar(&config.https, "https", false, "Use HTTPS to connect to lakeFS")
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+
+	if !config.enabled {
+		// system tests disabled
+		os.Exit(0)
+	}
+
+	// initialize the env/repo
+	logger = logging.Default()
+	logger.WithField("bucketPath", config.rawBucketPath).Info("Starting nessie run")
+
+	schemas := []string{"http"}
+	if config.https {
+		schemas = append(schemas, "https")
+	}
+	cl := genclient.NewHTTPClientWithConfig(strfmt.Default, &genclient.TransportConfig{
+		Host:     config.baseURL,
+		BasePath: genclient.DefaultBasePath,
+		Schemes:  schemas,
+	})
+
+	ctx, _ := context.WithTimeout(context.Background(), maxSetup)
+	for {
+		_, err := cl.HealthCheck.HealthCheck(nil)
+		if err == nil {
+			break
+		}
+
+		logger.WithError(err).Info("Setup failed")
+
+		select {
+		case <-ctx.Done():
+			panic("setup failed after all retries")
+		case <-time.After(5 * time.Second):
+		}
+	}
```

Comment:
Now `ctx` is somewhat poisoned, it will have a low timeout.  So if used downstream it may blow up occasionally.  Could we either scope it to here or make it `nil` or give it a more suggestive name like `setupCtx`?